tensorflow/core/kernels/conv_ops_gpu_3.cu.cc

```diff
@@ -422,29 +595,131 @@ void RunSwapDimension1And2InTensor3(const GPUDevice& d, const T* input,
+                      input_dims[2] >= kMinDimensionToUseRectTiles);
+  if (large_matrix) {
     // We get best performance when TileSize is the number of threads in a warp
     // (32 on our GPUs) and NumSubTiles is 8, so our block size is 8 * 32 = 256
     // threads.
     static const int TileSize = 32;
-    static const int NumSubTiles = 8;
+    static const int ThreadNum = 256;
+
     Dimension<3> input_dims_in_tiles = {
         input_dims[0], (input_dims[1] + TileSize - 1) / TileSize,
         (input_dims[2] + TileSize - 1) / TileSize,
     };
+
+    int total_tiles_count = input_dims_in_tiles[0] * input_dims_in_tiles[1] *
+                            input_dims_in_tiles[2];
+    SwapDimension1And2InTensor3UsingTiles<
+        T, ThreadNum, TileSize,
+        TileSize><<<total_tiles_count, ThreadNum, 0, d.stream()>>>(
+        input, input_dims, output);
+
+  } else if (narrow_matrix) {
+    // Define available tile sizes here for each size of data type supported:
+    std::map<int, int> tile_spec_128 = {{32, 15}, {64, 15}, {128, 15},
+                                        {256, 2}};
+    std::map<int, int> tile_spec_64  = {{32, 15},  {64, 15}, {128, 15},
+                                        {256, 8},  {512, 2}};
+    std::map<int, int> tile_spec_32  = {{32, 15},  {64, 15}, {128, 15},
+                                        {256, 10}, {512, 4}, {1024, 2}};
+    std::map<int, int> tile_spec_16  = {{32, 15},  {64, 15}, {128, 15},
+                                        {256, 10}, {512, 4}, {1024, 2}};
+    std::map<int, int> tile_spec_8   = {{32, 15},  {64, 15}, {128, 15},
+                                        {256, 10}, {512, 4}, {1024, 2}};
+
+    // Organize these tile size specifications into a map that maps from data
+    // type sizes to their specifications.
+    std::map<int, std::map<int, int>> tile_spec_map = {{128, tile_spec_128},
+                                                       {64, tile_spec_64},
+                                                       {32, tile_spec_32},
+                                                       {16, tile_spec_16},
+                                                       {8, tile_spec_8}};
+
+    std::map<int, int> tile_spec = tile_spec_map[8 * sizeof(T)];
+
+    int tile_long_side_len = 0;
+    int tile_short_side_len = 0;
+    float lowest_cost = std::numeric_limits<float>::max();
+    int data_long_side = max(input_dims[1], input_dims[2]);
+
+    for (std::map<int, int>::iterator it = tile_spec.begin();
+         it != tile_spec.end(); ++it) {
+      int proposed_tile_long_side_len = it->first;
+
+      // Threads that will not be doing anything useful when reading the matrix
+      // because the thread block size is bigger than the data block size.
+      float wasted_threads = (data_long_side -
+                              data_long_side / proposed_tile_long_side_len *
+                                  proposed_tile_long_side_len);
```

Comment:
All of the operands here appear to be `int`s, so does the LHS need to be a float?